libs/SalesforceSDK/src/com/salesforce/androidsdk/security/SalesforceKeyGenerator.java

```diff
@@ -153,33 +153,45 @@ private synchronized static String generateEncryptionKey(String name) {
         return encryptionKey;
     }
 
-    private synchronized static String generateUniqueId(String name, int length) {
+    private synchronized static String generateUniqueIdIfNoneStored(String name, int length) {
         final String id = readFromSharedPrefs(ID_PREFIX + name);
+        String uniqueId = null;
 
         // Checks if we have a unique identifier stored.
         if (id != null) {
             final PrivateKey privateKey = KeyStoreWrapper.getInstance().getRSAPrivateKey(KEYSTORE_ALIAS);
-            return Encryptor.decryptWithRSA(privateKey, id);
-        } else {
-            String uniqueId;
-            try {
-                // Create the key generator with its recommended secure random number generator provider algorithm.
-                final KeyGenerator keyGenerator = KeyGenerator.getInstance(AES);
-                keyGenerator.init(length);
-
-                // Generates a 256-bit key.
-                uniqueId = Base64.encodeToString(keyGenerator.generateKey().getEncoded(), Base64.NO_WRAP);
-            } catch (NoSuchAlgorithmException e) {
-                SalesforceSDKLogger.e(TAG, "Security exception thrown", e);
-
-                // Generates a random UUID 128-bit key instead.
-                uniqueId = UUID.randomUUID().toString();
+            uniqueId = Encryptor.decryptWithRSA(privateKey, id, Encryptor.CipherMode.RSA_PKCS1);
+            if (uniqueId == null) {
+                SalesforceSDKLogger.e(TAG, "Unable to decrpyt unique identifier stored");
             }
+        }
+
+        // Other create a new one and store it
+        if (uniqueId == null) {
```

Comment:
Previously if we failed to decrypt the stored unique id, we would return null which would cause the app to crash. Now it will recreate a unique id and log the problem encountered.

NB: That situation should not happen. I ran into it when I changed the cipher mode for all RSA calls. With this PR, only push notification encryption will use the new cipher mode. If/when we decide to change the cipher mode for the SalesforceKeyGenerator call, we will need to add a upgrade step.